File: leoz-node/src/main/java/org/deku/leoz/node/config/MessageListenerConfiguration.java
package org.deku.leoz.node.config;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.deku.leoz.messaging.activemq.ActiveMQContext;
import org.deku.leoz.node.App;
import org.deku.leoz.node.auth.Identity;
import org.deku.leoz.node.messaging.MessageListener;
import org.deku.leoz.node.messaging.auth.AuthorizationMessageHandler;
import org.deku.leoz.node.messaging.auth.v1.AuthorizationMessage;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Lazy;
import org.springframework.context.annotation.Profile;
import sx.jms.embedded.Broker;
import sx.jms.embedded.activemq.ActiveMQBroker;

import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;

/**
 * Message listener configuration.
 * Initializes message listener(s) and their message handlers
 * Created by masc on 20.06.15.
 */
@Profile( { App.PROFILE_CLIENT_NODE} )
@Configuration
@Lazy(false)
public class MessageListenerConfiguration {
    private Log mLog = LogFactory.getLog(MessageListenerConfiguration.class);

    MessageListener mMessageListener;

    /**
     * Broker event listener
     */
    Broker.EventListener mBrokerEventListener = new Broker.EventListener() {
        @Override
        public void onStart() {
            startIfReady();
        }

        @Override
        public void onStop() {
            stop();
        }
    };

    /**
     * Identity event listener
     */
    Identity.Listener mIdentityEventListener = new Identity.Listener() {
        @Override
        public void onIdUpdated(Identity identity) {
            startIfReady();
        }
    };

    /**
     * Indicates if message listener is ready to start (prerequisites are met)
     * @return
     */
    private boolean isReadyToStart() {
        return ActiveMQContext.instance().getBroker().isStarted() &&
                IdentityConfiguration.instance().getIdentity().getId() != null;
    }

    /**
     * Start message listener
     */
    private void startIfReady() {
        this.stop();

        if (this.isReadyToStart()) {
            // Configure and create listener
            mMessageListener = new MessageListener(
                    ActiveMQContext.instance(),
                    IdentityConfiguration.instance().getIdentity());

            // Add message handler delegatess
            mMessageListener.addDelegate(AuthorizationMessage.class, new AuthorizationMessageHandler());

            mMessageListener.start();
        }
    }

    private void stop() {
        if (mMessageListener != null) {
            mMessageListener.dispose();
            mMessageListener = null;
        }
    }

    @PostConstruct
    public void onInitialize() {
        mLog.info("Initializing node message listener");

        // Register event listeners
        ActiveMQBroker.instance().getDelegate().add(mBrokerEventListener);
        IdentityConfiguration.instance().getIdentity().getDelegate().add(mIdentityEventListener);

        this.startIfReady();
    }

    @PreDestroy
    public void onDestroy() {
        if (mMessageListener != null)
            mMessageListener.stop();
    }
}

File: leoz-node/src/main/java/org/deku/leoz/node/config/MessageListenerConfiguration.kt
package org.deku.leoz.node.config

import org.deku.leoz.config.ActiveMQConfiguration
import org.deku.leoz.node.Application
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Lazy
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Service
import sx.jms.Channel
import sx.jms.Broker
import sx.jms.activemq.ActiveMQBroker
import sx.jms.listeners.SpringJmsListener
import java.util.concurrent.ExecutorService
import javax.annotation.PostConstruct
import javax.annotation.PreDestroy
import javax.inject.Inject

/**
 * Message listener configuration.
 * Initializes message listener(s) and their message handlers
 * Created by masc on 20.06.15.
 */
@Profile(Application.PROFILE_CLIENT_NODE)
@Configuration
@Service
@Lazy(false)
open class MessageListenerConfiguration {
    private val log = LoggerFactory.getLogger(MessageListenerConfiguration::class.java)

    @Inject
    private lateinit var application: Application
    @Inject
    private lateinit var executorService: ExecutorService

    // Listeners
    val nodeQueueListener by lazy {
        SpringJmsListener(
                { Channel(ActiveMQConfiguration.instance.nodeQueue(this.application.identity.key)) },
                executorService)
    }

    val nodeNotificationListener by lazy {
        SpringJmsListener(
                { Channel(ActiveMQConfiguration.instance.nodeNotificationTopic) },
                executorService)
    }

    //region Lifecycle
    @PostConstruct
    open fun onInitialize() {
        // Register event listeners
        ActiveMQBroker.instance.delegate.add(brokerEventListener)
        this.startIfReady()
    }

    @PreDestroy
    open fun onDestroy() {
        this.stop()
    }

    /**
     * Broker event listener
     */
    private var brokerEventListener: Broker.EventListener = object : Broker.DefaultEventListener() {
        override fun onStart() {
            startIfReady()
        }

        override fun onStop() {
            stop()
        }
    }

    /**
     * Start message listener
     */
    @Synchronized private fun startIfReady() {
        this.stop()

        if (ActiveMQConfiguration.instance.broker.isStarted) {
            this.nodeQueueListener.start()
            this.nodeNotificationListener.start()
        }
    }

    /**
     * Stop message listener
     */
    @Synchronized private fun stop() {
        this.nodeQueueListener.stop()
        this.nodeNotificationListener.stop()
    }
    //endregion
}
